### DISCUSSION TITLE
Global slash commands not working?

### DISCUSSION THREAD

**Discussion by Kooloog** [Rapptz/discord.py#8967] [Q&A] [answered]
Hello, I'm trying to deploy an app on Heroku, enabling slash commands for all servers. When testing, I did set the guild argument to my server id and it worked just fine. However, when I tried removing such argument (in order to make my slash commands work in all servers), I waited for about 12 hours and Discord still says my bot doesn't have slash commands implemented, and of course they don't work in any server, why's that?

Here's how I've got everything set up:

```
import asyncio
import os
import boto3
import discord

from discord import app_commands

import triviaesp


class Client(discord.Client):
    async def on_ready(self):
        await self.wait_until_ready()
        await tree.sync()


client = Client(intents=discord.Intents.default())
tree = app_commands.CommandTree(client)


@client.event
async def on_ready():
    await client.change_presence(activity=discord.Game(name="Use the /help command!"))


# HELP COMMAND

@tree.command(name="help", description="[ENG] Show command help")
async def helpcommand(ctx, arg1: str = "def"):
    # Command implementation. Is tested and will work when slash commands work.

# REST OF COMMANDS
# ...

client.run(token)
```

Here are all permissions I enabled from the Discord Developer Portal: https://i.imgur.com/k0m7DrL.png

And here's what Discord has to say about my bot not using slash commands: https://i.imgur.com/D7CIghL.png

Any help is appreciated, thanks in advance!

**Comment by LostLuma** [comment, marked as answer]
You're overwriting the `on_ready` handler defined your `Client` class when you define the event again and add it via the `@event` decorator further below, as the library only supports having one of each event handler via the `Client` class.

You'll want to either combine the code inside the two functions, or sync your commands inside the [`setup_hook`](https://discordpy.readthedocs.io/en/stable/api.html#discord.Client.setup_hook) function instead.

**Reply by Kooloog** [reply]
Hello and thanks for answering! I found a handy example by Rapptz on this repo, and it matches your answer. Here's my new code after changing it, in case there's still something that seems out of place:

```
class Client(discord.Client):
    def __init__(self, *, intents: discord.Intents):
        super().__init__(intents=intents)
        self.tree = app_commands.CommandTree(self)

    async def setup_hook(self):
        await self.tree.sync()


client = Client(intents=discord.Intents.default())


@client.event
async def on_ready():
    await client.change_presence(activity=discord.Game(name="Use the /help command!"))
```

Now, Discord DOES say my bot has slash commands. They still don't work on servers, but I'm guessing it's because I still gotta wait a few hours before they're fully operational. I'll set your reply as answer when I see it working. Thanks a lot!

EDIT: It works!

**Reply by tailoric** [reply]
Just a side note.
Don't set your presence in `on_ready` the Client constructor has an [activity argument](https://discordpy.readthedocs.io/en/latest/api.html?highlight=discord%20client#discord.Client.activity) that you can use to set your activity when going online.
just set the activity in the `super().__init__()`
